### ISSUE TITLE
Set up automated periodic updates of external datasources

### ISSUE THREAD

**Issue by tskir** [EBIvariation/trait-curation#34]
Related to #32. In addition to being able to trigger imports manually, there should also be automated periodic updates.

Care needs to be taken (in both cases) to avoid any race conditions when a user tries to change something in the same moment when an update operation is being performed.

**Comment by joj0s** [comment]
@tcezard  I can confirm task automation is successful. What tasks should I schedule at this point and when? Available background tasks right now are ZOOMA suggestion imports and ClinVar imports, plus OLS imports when PR #77 is merged.

**Comment by joj0s** [comment]
Also, there exists a very useful Django package that lets the user schedule tasks through the Django Admin with a very friendly interface.
![image](https://user-images.githubusercontent.com/36301625/90748395-d0ad7d00-e2da-11ea-9d6d-c06e5593a7bf.png)
So maybe it would be preferable that I don't hardcode any scheduled task, but I let the admins do that?

**Comment by tcezard** [comment]
It would make a lot more sense not to have to reinvent the wheel and use this module. Can we predefine the tasks ? Does it run python function or command line executable ?

**Comment by tskir** [comment]
> By using the mentioned django_celery_beat module, we can configure the hardcoded task schedule via the Django Admin, which will save us the time of creating our own configuration page for that. However if the server restarts, the hardcoded schedule will again be used.

I'd say this solution is not very good, as it lacks settings persistence.

> The alternative, while still using the module, is to not hardcode anything, but instead have the admins define the scheduled tasks via the Django Admin. Again, the tasks will be already defined, and all the admins will have to do, is to select the task to run from a dropdown menu, and when to run them via a form.

This sounds somewhat better, but would it also just reset after every server restart?

Overall, I'd say that while it would be nice to have some sort of a UI to configure periodic import schedule, at this point this is not completely necessary. This is a setting which will be changed very very rarely and only by the app's admins. So even hardcoding that in settings.py sounds acceptable to me, if there are no other good (persistent) solutions

**Comment by tskir** [comment]
Also I wonder if we could use the admin page in issue #50 for that. There, we already have some configurable parameters, which will have to be persistent. For example, we have a configurable ZOOMA base URL. Could we just add a conifgurable ZOOMA query schedule field next to that? Or does it _have_ to be hardcoded in settings.py?

**Comment by joj0s** [comment]
> This sounds somewhat better, but would it also just reset after every server restart?

It would not. Since no hardcoded scheduling will exist when the server starts, the admin defined periodic tasks will just be saved in the database and they will persist when the server restarts. This is what it look like right now in the admin home page:
![image](https://user-images.githubusercontent.com/36301625/91436861-9e1ef980-e871-11ea-9563-255daae47cb4.png)

This is what the `Periodic Tasks` entry looks like:
![image](https://user-images.githubusercontent.com/36301625/91436945-b858d780-e871-11ea-9e7f-49024606e51d.png)

And this is what the form for adding a periodic task looks like:
![image](https://user-images.githubusercontent.com/36301625/91437159-18e81480-e872-11ea-9ed8-536616868099.png)

Basically, you choose from the set of hardcoded tasks, and you define a schedule for them

**Comment by tskir** [comment]
Right, I see. This looks reasonable. Now, the only drawback is that, after each deployment, the admins of the app would presumably have to configure those scheduled tasks from scratch. I wonder if there is a way to register some of those tasks in that plugin by default, so that even a blank deployment would contain them already?

**Comment by joj0s** [comment]
@tskir So we decided to hardcode the task schedules, however, celery still needs to produce an external file to read the hardcoded schedules from. Given that, I think the best option is to use the above mentioned `django-celery-beat` plugin to store the schedules in the database instead of a separate file, and have them available to the admins to change if need be.

**Comment by tskir** [comment]
@joj0s Yep, sounds reasonable to me, I agree with this approach
